src/veins/base/toolbox/SignalUtils.cc

```diff
@@ -22,264 +22,108 @@
-        }
-
-        auto next = std::next(it);
-        if (next == changes.end() || it->time != next->time) {
-            if (interference < minimum) minimum = interference;
+    // extract valid signals on the channel at the time of interest
+    // TODO: possibly move this filtering outside of this function
+    std::vector<Signal*> interferers;
+    for (auto& interfererFrame : interfererFrames) {
+        Signal* interferer = &interfererFrame->getSignal();
+        if (interferer->getReceptionStart() <= now && interferer->getReceptionEnd() > now && interfererFrame != exclude) {
+            interferers.push_back(interferer);
         }
-        it++;
     }
 
-    return minimum;
-}
-
-/*
- * Assume everything is at maximum value at the beginning (like no AMs applied)
- * Then apply AMs one after the other
- * If value is already below threshold before all AMs are applied, abort and return true
- */
-bool smallerAtFreqIndex(simtime_t start, simtime_t end, AirFrameVector& airFrames, size_t freqIndex, double threshold, AirFrame* exclude)
-{
-    ASSERT(start == end);
-
-    // Assume that threshold is >0 -> if there is no other AirFrame, this is 0 -> return true (0 < threshold)
-    if (airFrames.empty()) return true;
-
-    auto changes = calculateChanges(start, end, airFrames, exclude);
-    std::sort(changes.begin(), changes.end(), compareByTime);
-
-    // Works fine so far, as there is at least one AirFrame
-    if (changes.empty()) return true;
-
-    // Assumption: There is no AM with an attenuation > 1
-    uint16_t maxAnalogueModels = airFrames.front()->getSignal().getNumAnalogueModels();
-    for (uint16_t i = 0; i <= maxAnalogueModels; i++) {
-        auto it = changes.begin();
-
-        double channelLoad = 0;
-
-        while (it != changes.end()) {
-            if (it->time > start) break;
-
-            channelLoad += (*(it->signal))[freqIndex];
-
-            it++;
-        }
-
-        double minimum = channelLoad;
-
-        // Calculate all chunks
-        while (it != changes.end()) {
-            if (it->type == ChangeType::starting) {
-                channelLoad += (*(it->signal))[freqIndex];
-            }
-            else if (it->type == ChangeType::ending) {
-                channelLoad -= (*(it->signal))[freqIndex];
-            }
+    // check once before applying analogModels
+    if (powerLevelSumAtFrequencyIndex(interferers, freqIndex) < threshold) {
+        return true;
+    }
 
-            auto next = std::next(it);
-            if (next == changes.end() || it->time != next->time) {
-                if (channelLoad < minimum) minimum = channelLoad;
-            }
-            it++;
+    // start applying analogue models
+    auto analogueModelCount = interfererFrames.front()->getSignal().getAnalogueModelList()->size();
+    for (size_t analogueModelIndex = 0; analogueModelIndex < analogueModelCount; ++analogueModelIndex) {
+        for (auto signalPtr : interferers) {
```

Comment:
Maybe std::for_each?

Reply:
This is unfortunately less readable since the range specification and lambda definition/bind adds lots of boilerplate.